ttnn/cpp/ttnn/tensor/shape/shape_base.cpp

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "shape_base.hpp"
+#include <stdexcept>
+#include "fmt/color.h"
+#include "tt_metal/common/assert.hpp"
+
+namespace ttnn {
+
+namespace {
+
+int32_t normalized_index(int32_t index, size_t original_size, size_t container_size) {
+    int32_t orig_size = static_cast<int32_t>(original_size);
+    int32_t full_size = static_cast<int32_t>(container_size);
+
+    int fixed_index = index;
+    if (fixed_index < 0) {
+        fixed_index += full_size;
+    } else {
+        fixed_index += full_size - orig_size;
+    }
+
+    if (fixed_index < 0 || fixed_index >= full_size) {
+        TT_THROW("ShapeBase[] index out of range. {} not in [{}, {})", index, -full_size, full_size);
+    }
+
+    return fixed_index;
+}
+}
+
+void ShapeBase::init() {
+    m_original_size = m_value.size();
+    const size_t min_internal_size = 4;
+
+    if(m_original_size < min_internal_size) {
+        Container ones(min_internal_size - m_original_size, 1);
+        m_value.insert(m_value.begin(), ones.begin(), ones.end());
+    }
+}
```

Comment:
Shape will always be at least 4D? Is this necessary?

Reply:
This is bridge with a whole lot of code like shape[-1] and shape[-2] which assume that there are no tensors with rank 0 or that tile tensors always have rank >= 2.  

So we internal representation is always 4D+.
But for user - it is whatever they asked for.